### PR TITLE
DAOS-12601 mercury: update to 2.3.0rc5

### PR DESCRIPTION
**Pull request by soumagne**
Updates to latest mercury that enables multi-recv.

Drop previous na_ucx.c patch as change is now merged upstream.

Update RPM build and spec to make use of NA dynamic plugins for both OFI and UCX

Fix source URL, use bz2 archive that includes submodules and package perf tests as part of RPM

Remove libpsm_infinipath1 and libffi7 dependencies

Make use of RPM macros when possible

### CODE REVIEW

**Comment by soumagne** [comment]
Additional testing is disabled as part of that PR because it requires DAOS changes for both 2.2 and master.

**Review by frostedcmos — APPROVED** [review]

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
This is a really super cleanup of this RPM.  Really love to see the UCX bits being handled by a plugin now rather than the hacky double-build we used to have to do.

**Comment by brianjmurrell** [comment]
> Additional testing is disabled as part of that PR because it requires DAOS changes for both 2.2 and master.

Is that the change from the `daosio` set of tests to `pr` tests or this referring to the use of:

```
Skip-unit-tests: true
Skip-test: true
```
For this PR?

The need for the above test skipping is telling us there is a problem with this PR.  That testing that you are skipping is precisely meant to tell us if landing this PR will break existing DAOS builds, such as PRs in-flight for example.

> DAOS PRs are: [daos-stack/daos#11715](https://github.com/daos-stack/daos/pull/11715) (2.2) and [daos-stack/daos#11616](https://github.com/daos-stack/daos/pull/11616) (master)

Can the above two land without this PR and use the existing mercury?  If so, then it seems the order should be to land those first, require everyone to rebase/merge the above into their PRs, then give CI time to drain of in-flight PRs, then land the mercury update.

If not, that's a problem as we cannot have (or at least like to avoid) "flag day" events such as these where we need to land two things together that requires that everybody update and also is going to cause failures in any PRs that are already in CI due to having a non-compatible DAOS and mercury (once this mercury PR lands and becomes the mercury that all in-flight PRs will use).

File: Jenkinsfile [hunk 1]
```diff
@@ -44,4 +44,4 @@
 packageBuildingPipelineDAOSTest(['distros': ['el8', 'centos7', 'leap15', 'ubuntu20.04'],
                                  'make args': 'DISTRO_VERSION_EL8=8.5',
                                  'rpmlint_rpms_allow_errors': true,
-                                 'test-tag': 'daosio'])
+                                 'test-tag': 'pr'])
```
[the remark this answers is not in the document]
Comment (soumagne) [reply Jenkinsfile:47]: I was not entirely sure of what should be used, it sounded like `daosio` was not the right tag, @frostedcmos suggested to use `pr`, I'm just trying to target something that is not too heavy in terms of testing but still covers enough stuff. To be fair we should create a special mercury tag but that needs to be defined too.

File: debian/rules [hunk 2]
```diff
@@ -15,21 +15,23 @@ export DH_VERBOSE = 1
...
+	-DBUILD_TESTING:BOOL=OFF                                 \
+	-DMERCURY_ENABLE_DEBUG:BOOL=ON                           \
+	-DMERCURY_INSTALL_LIB_DIR:PATH=lib/$(DEB_HOST_MULTIARCH) \
+	-DMERCURY_USE_BOOST_PP:BOOL=ON                           \
+	-DMERCURY_USE_CHECKSUMS:BOOL=OFF                         \
+	-DMERCURY_USE_SYSTEM_BOOST:BOOL=ON                       \
+	-DMERCURY_USE_XDR:BOOL=OFF                               \
+	-DNA_USE_SM:BOOL=ON                                      \
+	-DNA_USE_OFI:BOOL=ON                                     \
+    $(NULL)
```
[the remark this answers is not in the document]
Comment (soumagne) [reply debian/rules:34]: thanks for catching that, will fix it.

File: mercury.spec [hunk 3]
```diff
@@ -20,141 +19,93 @@ Release: 6%{?dist}
...
+        -DMERCURY_USE_XDR:BOOL=OFF                  \
+        -DNA_USE_DYNAMIC_PLUGINS:BOOL=ON            \
+        -DNA_DEFAULT_PLUGIN_PATH:PATH=%{_libdir}    \
+        -DNA_USE_SM:BOOL=ON                         \
+        -DNA_USE_UCX:BOOL=%{ucx}                    \
+        -DNA_USE_OFI:BOOL=ON
+
+%{?cmake_build}
+# necessary for old cmake environments (e.g., CentOS7)
+%{!?cmake_build:%__cmake --build %{_vpath_srcdir}}
```
[the remark this answers is not in the document]
Comment (soumagne) [reply mercury.spec:103]: ok I can try to arrange that a bit.

File: mercury.spec [hunk 4]
```diff
@@ -168,25 +119,33 @@ rm -rf $RPM_BUILD_ROOT/.variants
 
 %files
 %license LICENSE.txt
-%exclude %{_libdir}/*.so.*.ucx
+%{_bindir}/hg_*
+%{_bindir}/na_*
 %{_libdir}/*.so.*
-%doc
+%{_libdir}/libna_plugin_ofi.so
```
[the remark this answers is not in the document]
Comment (soumagne) [reply mercury.spec:125]: thanks I appreciate the pointer, I was actually wondering what was the right guideline for it. I will move it to a separate directory, we do have the cmake build option already to support that.

**Comment by soumagne** [comment]
Right I was referring to the `skip-test` part. Yes landing that will break existing PRs... the problem is that I had to introduce an API change in the info struct that we use to initialize mercury so that we can keep track of the version of the struct that is being used and prevent further breakage of that sort... I realize that is not ideal though it will fix further issues. If we really want to avoid PR breakage, we could temporarily pin DAOS to not pick up any newer mercury until all mercury PRs are landed ? is it an option ?

File: Jenkinsfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply Jenkinsfile:47]: Indeed, I am not entirely sure either.  I am just skeptical that `pr` is the right choice.  Frankly I don't think `pr` is probably the right choice for just about any PR going into CI.

Historically the set of tests were tagged `pr` was at one time actually stripped back by quite a bit due to the load the `pr` set was putting on our clusters at the time.  When it was stripped back, it was also expected that developers would add `Features:` tags to choose a group of tests that would target the code changes in their PRs.  Unfortunately I don't think too many developers actually do this.  There is work underway to try to somewhat heuristically suggest tags at commit time so hopefully things will get better there.

So, ideally here, want a set of test tags that specifically exercises mercury, which I guess means tests that focus on network IO (realizing of course that anything to do with DAOS involves network IO), and perhaps stress tests it and/or pushes into the corner cases of network IO, etc.

I don't know which tags that is, and as you suggest it would be useful to curate a set of tests and tag them mercury.  Indeed, I have suggested in the past that this be done for every one of our dependencies so that we can imply the dependency name into the test tag to be used.

Maybe a query to the Functional test engineers about which tests would specifically focus on networking and mercury might get you some input.  I don't know what the `daosio` tests are but the tag name sounds good, in terms of targeting network IO.  I suppose that could be disk IO also though.  :smirk: 

It looks like `daosio` was added by @jgmoore-or in https://github.com/daos-stack/mercury/commit/8c5dc136833322c8eff565df0caba28e314d8867.  Maybe he has some insight into choosing that tag specifically.

**Comment by soumagne** [comment]
I'm going to try to see if there's a workaround for it. It's not an API change in the sense that the actual API functions changed but yes that struct changed and it can create an ABI change, I agree that in theory that should qualify as a bump in the SOVERSION. I'd just try to avoid bumping the version just for that change...

**Comment by brianjmurrell** [comment]
> soumagne force-pushed the soumagne/230_update branch 2 times, most recently from 08c4f79 to d0219b6 
> soumagne force-pushed the soumagne/230_update branch from d0219b6 to 3cd9181 

Please don't forget to refrain from force-pushing after reviewing has started.

**Comment by soumagne** [comment]
oh well I was actually trying to make it nice and clean for you to review by having a single separate commit that addresses your comments from the origin PR commit :)

**Comment by brianjmurrell** [comment]
> oh well I was actually trying to make it nice and clean for you to review by having a single separate commit that addresses your comments from the origin PR commit :) see [3cd9181](https://github.com/daos-stack/mercury/commit/3cd918185712dc37b1ba16e8c17d1fff3745ecd6)

So, one thing you can do is force-push to the most recently reviewed commit.  So that is, if you pushed commits 1-4 and then requested reviews, and got reviews on them, and then pushed commits 5-8 to address review comments, you can squash 5-8 into a single commit and force push that before re-requesting reviews.  That will make all of the new commits a single commit but still preserve the state of the reviewer's previous reviews.

But generally speaking, the number of commits in a PR has little-to-no bearing on the difficultly level of reviewing the PR as several (or usually, or at least initially, all) commits are reviewed together as a group without any knowledge even of how many commits are being reviewed.

I would much much prefer to have my previous review state preserved over simply reducing the number of commits in the PR.

**Comment by soumagne** [comment]
understood, will make sure I follow that in the future. Waiting now on https://github.com/daos-stack/daos/pull/11948 and https://github.com/daos-stack/daos/pull/11947 to be merged before re-enabling the testing.

**Review by jgmoore-or — DISMISSED** [review]

File: Jenkinsfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (soumagne) [reply Jenkinsfile]: I reverted back to `daosio`.

**Review by frostedcmos — APPROVED** [review]

**Review by jgmoore-or — APPROVED** [review]

**Review by jolivier23 — DISMISSED** [review]

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
Really nice to see the big improvement in the building of the optional `-ucx` subpackage.

Also nice job cleaning this spec up.

Might also be nice to update the dates in the changelogs since you have to push the new patch to the PR and fix the mentioned `\` alignment.

File: debian/rules [hunk 5]
```diff
@@ -15,21 +15,23 @@ export DH_VERBOSE = 1
...
-	-DMERCURY_USE_BOOST_PP=ON \
-	-DMERCURY_USE_SYSTEM_BOOST=ON \
-	-DMERCURY_ENABLE_VERBOSE_ERROR=ON \
-	-DBUILD_TESTING=OFF \
-	-DNA_USE_OFI=ON \
-	-DBUILD_DOCUMENTATION=OFF \
-	-DMERCURY_INSTALL_LIB_DIR=lib/$(DEB_HOST_MULTIARCH) \
-	-DBUILD_SHARED_LIBS=ON  \
-         $(NULL)
+cmake_options =                                              \
```
[the remark this answers is not in the document]
Comment (soumagne) [reply debian/rules:18]: ok I had to remove the tabs and replace with spaces, fixed now

**Comment by soumagne** [comment]
@brianjmurrell thanks I have addressed all this and updated the dates :)

**Review by frostedcmos — APPROVED** [review]

**Review by brianjmurrell — APPROVED** [review]
Looks great.  I really do appreciate the effort you are putting into cleaning up and fleshing out these specfiles.  When one has lots to get through, one tends to be more brief about the details with the hope of having the time to go back and be more detailed but that followup tends to get overridden with a new #1 priority before it can be achieved.  :-(  It's nice that a component-owner takes the time to flesh out the details in their component's packaging.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
Oh wait!  I just noticed that there are some PIC warnings  (not really sure why these are warnings and no errors which would have made them fail the run) on CentOS 7:

```
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/hg_bw_read
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/hg_bw_write
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/hg_perf_server
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/hg_rate
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/na_bw_get
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/na_bw_put
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/na_lat
mercury.x86_64: W: position-independent-executable-suggested /usr/bin/na_perf_server
```

That's an SDL issue.  We should be shipping PIC executables.  Clearly EL8 and Leap15 are more advanced in that they are building PIC (generally hardened) binaries more implicitly.

While fixing that it would be nice to add exemptions for the missing manpages and maybe file a ticket upstream to get some manpages written/included:
```
mercury.x86_64: W: no-manual-page-for-binary hg_bw_read
mercury.x86_64: W: no-manual-page-for-binary hg_bw_write
mercury.x86_64: W: no-manual-page-for-binary hg_perf_server
mercury.x86_64: W: no-manual-page-for-binary hg_rate
mercury.x86_64: W: no-manual-page-for-binary na_bw_get
mercury.x86_64: W: no-manual-page-for-binary na_bw_put
mercury.x86_64: W: no-manual-page-for-binary na_lat
mercury.x86_64: W: no-manual-page-for-binary na_perf_server
```

As well as adding some documentation that is missing in the main package:

```
mercury.x86_64: W: no-documentation
```

Apologies that I missed these before.  One of the problems of having silent (to the final result) warnings/failures.  I do want to make these failures in the future but wanted a window of time where they would be warnings so that people can get fixes into their packages before they start becoming fatal failures.

**Review by jgmoore-or — DISMISSED** [review]

**Comment by soumagne** [comment]
@brianjmurrell I can't really fix those on centos 7 because the cmake version is too ancient to support PIE properly (this has nothing to do with PIC which is supported). These errors do not appear on el8 / leap15 builds because I was able to fix them there :) it was not implicitly done. If not supporting PIE is a concern, I can remove the executables from centos 7 builds. Could you please tell how to add exemptions? I forgot how to do that...

**Comment by brianjmurrell** [comment]
> @brianjmurrell I can't really fix those on centos 7 because the cmake version is too ancient to support PIE properly (this has nothing to do with PIC which is supported, see https://cmake.org/cmake/help/latest/policy/CMP0083.html#policy:CMP0083).

I see.  This puts us in a bit of a pickle then I think.

> These errors do not appear on el8 / leap15 builds because I was able to fix them there :) it was not implicitly done.

Ahh.  That's great then.  Thanks for that!

> If not supporting PIE is a concern, I can remove the executables from centos 7 builds.

The other option is to simply not build at all on CentOS 7 -- which we only support with 2.2 and less.  2.4 drops support for CentOS 7, so if we will never need this newer version of mercury on < 2.4, I would think we could simply stop building it there.  That is controlled in the `Jenkinsfile` as I am sure you will recall.

If we do need to keep supporting CentOS 7 with this version of mercury I guess we will need to consult with our security/SDL champion to see what our options/obligations are there.

> Could you please tell me how to add exemptions? I forgot how to do that...

You would need to add a `mercury.rpmlintrc` file.  An example of what that file looks like could be found in the [`libfabric.rpmlintrc`](https://github.com/daos-stack/libfabric/blob/master/libfabric.rpmlintrc).

**Comment by soumagne** [comment]
@brianjmurrell thanks for the comments, have a look again, I have added the rpmlintrc file with the filters. I ended up just not building the perf binaries on centos 7, it's optional to have them, that does not impact any library feature. I added doc for the main package but I don't have any specific doc for the ucx plugin.

File: Jenkinsfile [hunk 6]
```diff
@@ -43,5 +43,4 @@
 
 packageBuildingPipelineDAOSTest(['distros': ['el8', 'centos7', 'leap15', 'ubuntu20.04'],
                                  'make args': 'DISTRO_VERSION_EL8=8.5',
-                                 'rpmlint_rpms_allow_errors': true,
```
Comment (brianjmurrell) [inline Jenkinsfile:46]: Thanks for catching this.  We generally never want to use this option.  It's a huge hammer for very particular cases.

File: mercury.spec [hunk 7]
```diff
@@ -19,6 +19,14 @@ Release: 1%{?dist}
 %global ucx 1
 %endif
 
+# do not build perf binaries on CentOS7 due to CMake PIE issues
+# see: https://cmake.org/cmake/help/latest/policy/CMP0083.html#policy:CMP0083
+%if 0%{?rhel} >= 8 || 0%{?suse_version} >= 1315
+%global __build_perf 1
+%else
+%global __build_perf 0
+%endif
```
Comment (brianjmurrell) [inline mercury.spec:28]: No need to change this on this go around, but for future reference, this would be better as an exclude list rather than include list so that new platforms being added default to the default/desired behaviour and only specifically identified platforms use the abnormal behaviour.

IOW, we should only be setting this to `0` for centos7 as a known problem platform that needs it and new platforms that get ported will get a `1` by default.

**Review by brianjmurrell — APPROVED** [review]

**Review by frostedcmos — APPROVED** [review]

File: mercury.spec (hunk 7 above)
Comment (soumagne) [reply mercury.spec:28]: OK yes you're right I'll make a note to change that next time

**Review by jolivier23 — APPROVED** [review]

**Review by jgmoore-or — APPROVED** [review]

**Comment by soumagne** [comment]
All tests passed, please merge when you can.
